File: public/RingGeometry.js




THREE.RingGeometry = function ( innerRadius, outerRadius, thetaSegments, phiSegments, thetaStart, thetaLength ) {

    THREE.Geometry.call( this );

    innerRadius = innerRadius || 0;
    outerRadius = outerRadius || 50;

    thetaStart = thetaStart !== undefined ? thetaStart : 0;
    thetaLength = thetaLength !== undefined ? thetaLength : Math.PI * 2;

    thetaSegments = thetaSegments !== undefined ? Math.max( 3, thetaSegments ) : 8;
    phiSegments = phiSegments !== undefined ? Math.max( 3, phiSegments ) : 8;
    
    var i, o, uvs = [], radius = innerRadius, radiusStep = ( ( outerRadius - innerRadius ) / phiSegments);
   
    for( i = 0; i <= phiSegments; i++) {//concentric circles inside ring

        if( innerRadius === 0 && i === 0 ) {
            var vertex = new THREE.Vector3();
            this.vertices.push( vertex );
            uvs.push( new THREE.Vector2( ( vertex.x / radius + 1 ) / 2, - ( vertex.y / radius + 1 ) / 2 + 1 ) );
            continue;
        }
        
        for( o = 0; o <= thetaSegments; o++) {//number of segments per ring

            var vertex = new THREE.Vector3();
            
            vertex.x = radius * Math.cos( thetaStart + o / thetaSegments * thetaLength );
            vertex.y = radius * Math.sin( thetaStart + o / thetaSegments * thetaLength );
            
            this.vertices.push( vertex );
            uvs.push( new THREE.Vector2( ( vertex.x / radius + 1 ) / 2, - ( vertex.y / radius + 1 ) / 2 + 1 ) );
        }
        
        radius += radiusStep;

    }

    var n = new THREE.Vector3( 0, 0, 1 );
    
    for( i = 0; i < phiSegments; i++) {//concentric circles inside ring

        for( o = 0; o <= thetaSegments; o++) {//number of segments per ring
            var v1, v2, v3;

            if( innerRadius === 0 && i === 0 ) {
                v1 = 0;
                v2 = o+1;
                v3 = o+2;
                this.faces.push( new THREE.Face3( v1, v2, v3, [ n, n, n ] ) );
                this.faceVertexUvs[ 0 ].push( [ uvs[ v1 ], uvs[ v2 ], uvs[ v3 ] ]);
                continue;
            }

            v1 = o + (thetaSegments * i) + i;
            v3 = o + (thetaSegments * i) + thetaSegments + i;
            v2 = o + (thetaSegments * i) + thetaSegments + 1 + i;
            
            this.faces.push( new THREE.Face3( v1, v2, v3, [ n, n, n ] ) );
            this.faceVertexUvs[ 0 ].push( [ uvs[ v1 ], uvs[ v2 ], uvs[ v3 ] ]);
            
            v1 = o + (thetaSegments * i) + i;
            v3 = o + (thetaSegments * i) + 1 + i;
            v2 = o + (thetaSegments * i) + thetaSegments + 1 + i;
            
            this.faces.push( new THREE.Face3( v1, v2, v3, [ n, n, n ] ) );
            this.faceVertexUvs[ 0 ].push( [ uvs[ v1 ], uvs[ v2 ], uvs[ v3 ] ]);

        }
    }
    
/*    if(innerRadius === 0){
        var circ = new THREE.CircleGeometry(innerRadius + radiusStep, thetaSegments, thetaStart, thetaLength);
        this.vertices = this.vertices.concat(circ.vertices);
        this.faces = this.faces.concat(circ.faces);
        this.faceVertexUvs = this.faceVertexUvs.concat(circ.faceVertexUvs);
    }*/
//    this.mergeVertices();

    this.computeCentroids();
    this.computeFaceNormals();

    this.boundingSphere = new THREE.Sphere( new THREE.Vector3(), radius ); 

};

THREE.RingGeometry.prototype = Object.create( THREE.Geometry.prototype );

File: Assets/Scripts/RingGeometry.cs
/**
 * @author Kaleb Murphy
 */

using UnityEngine;
using System.Collections;
using System.Collections.Generic;

public class RingGeometry  {

	
	static public Mesh CreateRingGeometry (int innerRadius, int outerRadius, int thetaSegments, int phiSegments, int thetaStart, int thetaLength)
	{
		Mesh mesh = new Mesh ();
		
		List<Vector3> vertices = new List<Vector3> ();	
		List<int> triangles = new List<int> ();	
		List<Vector3> normals = new List<Vector3> ();	
		
		outerRadius = outerRadius == 0 ? outerRadius : 50;
		thetaLength = thetaLength == 0 ? thetaLength : 50;
		
		thetaSegments = thetaSegments == 0 ? Mathf.Max (3, thetaSegments) : 8;
		phiSegments = phiSegments == 0 ? Mathf.Max (3, phiSegments) : 8;
		
		int i, o, radius = innerRadius, radiusStep = ((outerRadius - innerRadius) / phiSegments);

		List<Vector2> uvs = new List<Vector2> (); 
		
		for (i = 0; i <= phiSegments; i++) {// concentric circles inside ring
			
			for (o = 0; o <= thetaSegments; o++) {// number of segments per circle
			
				Vector3 vertex = new Vector3 ();
				float segment = thetaStart + o / thetaSegments * thetaLength;
				
				vertex.x = radius * Mathf.Cos (segment);
				vertex.y = radius * Mathf.Sin (segment);
				
				vertices.Add (vertex);
				normals.Add (-Vector3.forward);
				
				uvs.Add (new Vector2 ((vertex.x / radius + 1) / 2, - (vertex.y / radius + 1) / 2 + 1));
			}
			
			radius += radiusStep;
			
		}
		
	
		
		for (i = 0; i < phiSegments; i++) {
			
			int thetaSegment = i * thetaSegments;
			
			for (o = 0; o <= thetaSegments; o++) {

				int segment = o + thetaSegment;
				
				int v1 = segment + i;
				int v2 = segment + thetaSegments + i;
				int v3 = segment + thetaSegments + 1 + i;
				
				triangles.Add (v1);
				triangles.Add (v2);
				triangles.Add (v3);
				
				v1 = segment + i;
				v2 = segment + thetaSegments + i + 1;
				v3 = segment + 1 + i;
				
				triangles.Add (v1);
				triangles.Add (v2);
				triangles.Add (v3);
				
			}
			
		}
		
		mesh.vertices = vertices.ToArray ();	
		mesh.triangles = triangles.ToArray ();
		mesh.normals = normals.ToArray ();
		mesh.uv = uvs.ToArray ();	
		
		return mesh;
		
	}
}
